File: backend/routes/image/proxy.ts
/**
 * backend/routes/image/proxy.ts
 * ─────────────────────────────────────────────────────────────────────────────
 * Image Proxy Route — Fetches external images on behalf of the frontend,
 * bypassing referrer restrictions from Webtoon / Manhwa CDNs.
 *
 * Features:
 *  • In-memory TTL cache (avoids repeated remote fetches for same URL)
 *  • ETag + Last-Modified / 304 Not Modified support
 *  • Content-Type validation (only allows image/* responses)
 *  • Response size guard (rejects payloads > MAX_PROXY_SIZE)
 *  • Configurable retry with exponential back-off
 *  • Dynamic Referer spoofing based on image URL origin
 *  • Colorized console logging with request latency
 *  • Served with 1-year CDN cache headers for hot assets
 * ─────────────────────────────────────────────────────────────────────────────
 */

import { Router, Request, Response } from 'express';
import { CacheStore } from '../../utils/cache.js';
import crypto from 'crypto';

const router = Router();

// ─── Config ──────────────────────────────────────────────────────────────────
const MAX_PROXY_SIZE_MB  = parseInt(process.env.MAX_PROXY_MB  || '20',  10);
const MAX_PROXY_SIZE     = MAX_PROXY_SIZE_MB * 1024 * 1024;
const PROXY_CACHE_TTL_MS = 30 * 60 * 1000;  // 30 minutes
const PROXY_MAX_RETRIES  = 3;
const PROXY_RETRY_BASE   = 400; // ms

// ─── In-memory proxy cache (separate from stitchedCache) ──────────────────────
interface ProxyCacheEntry {
  data:        Buffer;
  contentType: string;
  etag:        string;
  size:        number;
  fetchedAt:   number;
}

const proxyCache = new CacheStore<ProxyCacheEntry>(
  'proxyImages', PROXY_CACHE_TTL_MS, 300
);

// ─── Terminal colors (reuse pattern from server.ts) ─────────────────────────
const R  = '\x1b[0m';
const DIM = '\x1b[2m';
const CY  = '\x1b[36m';
const GN  = '\x1b[32m';
const YL  = '\x1b[33m';
const RD  = '\x1b[31m';
const BW  = '\x1b[97m';
const BM  = '\x1b[95m';
const label = (s: string) => `\x1b[1m${BM}${s}${R}`;
const ok    = (s: string) => `\x1b[1m${GN}${s}${R}`;
const warn  = (s: string) => `\x1b[1m${YL}${s}${R}`;
const err   = (s: string) => `\x1b[1m${RD}${s}${R}`;
const dim   = (s: string) => `${DIM}${s}${R}`;
const val   = (s: string) => `${BW}${s}${R}`;
const route = (s: string) => `${CY}${s}${R}`;

// ─── Helpers ─────────────────────────────────────────────────────────────────

/** Generate an MD5 fingerprint for a Buffer (used as ETag). */
function makeETag(buf: Buffer): string {
  return `"${crypto.createHash('md5').update(buf).digest('hex')}"`;
}

/** Derive a plausible Referer for CDN bypass based on the image URL. */
function spoofReferer(url: string): string {
  try {
    const parsed  = new URL(url);
    const host    = parsed.hostname.toLowerCase();
    if (host.includes('webtoons'))  return 'https://www.webtoons.com/';
    if (host.includes('naver'))     return 'https://comic.naver.com/';
    if (host.includes('kakao'))     return 'https://page.kakao.com/';
    if (host.includes('lezhin'))    return 'https://www.lezhin.com/';
    if (host.includes('tapas'))     return 'https://tapas.io/';
    if (host.includes('manhwa'))    return 'https://manhwatop.com/';
    return `${parsed.protocol}//${parsed.hostname}/`;
  } catch {
    return 'https://www.webtoons.com/';
  }
}

/** Fetch with exponential back-off retry on 5xx or network errors. */
async function fetchWithRetry(
  url: string,
  headers: Record<string, string>,
  retries = PROXY_MAX_RETRIES,
  baseDelay = PROXY_RETRY_BASE
): Promise<globalThis.Response> {
  let lastErr: unknown;
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      const resp = await fetch(url, { headers });
      // Only retry on 5xx
      if (resp.status >= 500 && attempt < retries - 1) {
        const delay = baseDelay * Math.pow(2, attempt);
        console.warn(
          `${label('[Proxy]')} ${warn(`Retry ${attempt + 1}/${retries}`)} ` +
          `${route(url.slice(0, 60))} → status ${resp.status} — waiting ${delay}ms`
        );
        await new Promise(r => setTimeout(r, delay));
        continue;
      }
      return resp;
    } catch (e: unknown) {
      lastErr = e;
      if (attempt < retries - 1) {
        const delay = baseDelay * Math.pow(2, attempt);
        console.warn(
          `${label('[Proxy]')} ${warn(`Network error, retry ${attempt + 1}/${retries}`)} ` +
          `— ${e.message} — waiting ${delay}ms`
        );
        await new Promise(r => setTimeout(r, delay));
      }
    }
  }
  throw lastErr || new Error('Max retries reached');
}

// ─── Route ───────────────────────────────────────────────────────────────────

router.get('/proxy-image', async (req: Request, res: Response) => {
  const start     = Date.now();
  const targetUrl = req.query.url as string;

  // ── Validate input ────────────────────────────────────────────────────────
  if (!targetUrl) {
    return res.status(400).json({ success: false, error: "Missing required query parameter: 'url'" });
  }

  // Unwrap any double-proxied URLs
  let fetchUrl = targetUrl;
  if (fetchUrl.includes('/api/proxy-image')) {
    const m = fetchUrl.match(/[?&]url=([^&]+)/);
    if (m?.[1]) fetchUrl = decodeURIComponent(m[1]);
  }

  // Validate URL scheme — only allow http/https
  try {
    const parsed = new URL(fetchUrl);
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return res.status(400).json({ success: false, error: 'Only HTTP/HTTPS URLs are supported.' });
    }
  } catch {
    return res.status(400).json({ success: false, error: 'Invalid URL format.' });
  }

  // ── Cache check ───────────────────────────────────────────────────────────
  const cacheKey = crypto.createHash('md5').update(fetchUrl).digest('hex');
  const cached   = proxyCache.get(cacheKey);

  if (cached) {
    // ETag conditional request (304 Not Modified)
    const clientETag = req.headers['if-none-match'];
    if (clientETag === cached.etag) {
      const ms = Date.now() - start;
      console.log(
        `${label('[Proxy]')} ${dim('304')} ${dim('CACHE HIT')} ` +
        `${route(fetchUrl.slice(0, 55))} ${dim(`(${ms}ms)`)}`
      );
      return res.status(304).end();
    }

    const ms = Date.now() - start;
    console.log(
      `${label('[Proxy]')} ${ok('200')} ${dim('CACHE HIT')} ` +
      `${route(fetchUrl.slice(0, 55))} ` +
      `${dim(`${(cached.size / 1024).toFixed(1)}KB`)} ${dim(`(${ms}ms)`)}`
    );

    res.setHeader('Content-Type',     cached.contentType);
    res.setHeader('ETag',             cached.etag);
    res.setHeader('X-Cache',          'HIT');
    res.setHeader('Cache-Control',    'public, max-age=31536000, immutable');
    res.setHeader('X-Proxy-Size-KB',  (cached.size / 1024).toFixed(1));
    return res.send(cached.data);
  }

  // ── Remote fetch ──────────────────────────────────────────────────────────
  try {
    const headers = {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      'Referer':    spoofReferer(fetchUrl),
      'Accept':     'image/webp,image/avif,image/jpeg,image/png,image/*,*/*;q=0.8',
      'Accept-Language': 'en-US,en;q=0.9',
      'Accept-Encoding': 'gzip, deflate, br',
      'Sec-Fetch-Dest': 'image',
      'Sec-Fetch-Mode': 'no-cors',
      'Sec-Fetch-Site': 'cross-site',
    };

    const response = await fetchWithRetry(fetchUrl, headers);

    if (!response.ok) {
      console.warn(
        `${label('[Proxy]')} ${warn(`Upstream error ${response.status}`)} ` +
        `${route(fetchUrl.slice(0, 60))}`
      );
      return res.status(response.status).json({
        success: false,
        error:   `Upstream returned ${response.status} ${response.statusText}`,
        url:     fetchUrl,
      });
    }

    // Validate Content-Type — reject non-image responses
    const contentType = response.headers.get('Content-Type') || 'image/jpeg';
    if (!contentType.startsWith('image/') && !contentType.includes('octet-stream')) {
      console.warn(
        `${label('[Proxy]')} ${warn('Blocked non-image response')} ` +
        `type=${val(contentType)} ${route(fetchUrl.slice(0, 50))}`
      );
      return res.status(415).json({
        success: false,
        error:   `Upstream returned non-image content type: ${contentType}`,
      });
    }

    // Guard response size
    const contentLength = parseInt(response.headers.get('content-length') || '0', 10);
    if (contentLength > MAX_PROXY_SIZE) {
      console.warn(
        `${label('[Proxy]')} ${warn('Blocked oversized response')} ` +
        `${(contentLength / 1024 / 1024).toFixed(1)}MB > ${MAX_PROXY_SIZE_MB}MB limit`
      );
      return res.status(413).json({
        success: false,
        error:   `Image exceeds maximum proxy size of ${MAX_PROXY_SIZE_MB}MB`,
      });
    }

    const arrayBuffer = await response.arrayBuffer();
    const buffer      = Buffer.from(arrayBuffer);

    // Post-download size check
    if (buffer.length > MAX_PROXY_SIZE) {
      return res.status(413).json({
        success: false,
        error:   `Downloaded image (${(buffer.length / 1024 / 1024).toFixed(1)}MB) exceeds limit`,
      });
    }

    const etag = makeETag(buffer);

    // Store in proxy cache
    proxyCache.set(cacheKey, {
      data:        buffer,
      contentType: contentType.split(';')[0].trim(),
      etag,
      size:        buffer.length,
      fetchedAt:   Date.now(),
    });

    const ms = Date.now() - start;
    console.log(
      `${label('[Proxy]')} ${ok('200')} ${dim('FETCH')} ` +
      `${route(fetchUrl.slice(0, 55))} ` +
      `${val(`${(buffer.length / 1024).toFixed(1)}KB`)} ${dim(`(${ms}ms)`)}`
    );

    res.setHeader('Content-Type',     contentType);
    res.setHeader('ETag',             etag);
    res.setHeader('X-Cache',          'MISS');
    res.setHeader('Cache-Control',    'public, max-age=31536000, immutable');
    res.setHeader('X-Proxy-Source',   new URL(fetchUrl).hostname);
    res.setHeader('X-Proxy-Size-KB',  (buffer.length / 1024).toFixed(1));
    return res.send(buffer);

  } catch (e: unknown) {
    const ms = Date.now() - start;
    console.error(
      `${label('[Proxy]')} ${err('ERROR')} ${route(fetchUrl.slice(0, 60))} ` +
      `— ${e.message} ${dim(`(${ms}ms)`)}`
    );
    return res.status(500).json({
      success: false,
      error:   `Proxy fetch failed: ${e.message}`,
      url:     fetchUrl,
    });
  }
});

// ── Cache stats endpoint ──────────────────────────────────────────────────────
router.get('/proxy-cache/stats', (_req: Request, res: Response) => {
  res.json({ success: true, ...proxyCache.stats() });
});

// ── Cache clear endpoint ──────────────────────────────────────────────────────
router.delete('/proxy-cache', (_req: Request, res: Response) => {
  const size = proxyCache.size;
  proxyCache.clear();
  console.log(`${label('[Proxy]')} ${warn('Cache cleared')} — removed ${size} entries`);
  res.json({ success: true, cleared: size });
});

export default router;

File: backend/python/routes/proxy.py
"""
backend/python/routes/proxy.py
─────────────────────────────────────────────────────────────────────────────
Image Proxy Route — Fetches external images on behalf of the frontend,
bypassing referrer restrictions from Webtoon / Manhwa CDNs.
─────────────────────────────────────────────────────────────────────────────
"""

import logging
import os
import json
import hashlib
import time
import httpx
import asyncio
from urllib.parse import urlparse, unquote, parse_qs
from fastapi import APIRouter, Request, Response, Query, HTTPException

from utils.cache import proxy_cache

logger = logging.getLogger("anivox.routes.proxy")
router = APIRouter()

# ─── Config ──────────────────────────────────────────────────────────────────
MAX_PROXY_SIZE_MB = int(os.getenv("MAX_PROXY_MB", "20"))
MAX_PROXY_SIZE = MAX_PROXY_SIZE_MB * 1024 * 1024
PROXY_CACHE_TTL_SEC = 30 * 60  # 30 minutes
PROXY_MAX_RETRIES = 3
PROXY_RETRY_BASE_SEC = 0.4

# ─── Helpers ─────────────────────────────────────────────────────────────────

def make_etag(buf: bytes) -> str:
    """Generate MD5 fingerprint for a bytes buffer (used as ETag)."""
    return f'"{hashlib.md5(buf).hexdigest()}"'


def spoof_referer(url: str) -> str:
    """Derive a plausible Referer for CDN bypass based on the image URL."""
    try:
        parsed = urlparse(url)
        host = (parsed.hostname or "").lower()
        if "webtoons" in host:
            return "https://www.webtoons.com/"
        if "naver" in host:
            return "https://comic.naver.com/"
        if "kakao" in host:
            return "https://page.kakao.com/"
        if "lezhin" in host:
            return "https://www.lezhin.com/"
        if "tapas" in host:
            return "https://tapas.io/"
        if "manhwatop" in host or "manhwa" in host:
            return "https://manhwatop.com/"
        if "manhuato" in host or "manhua" in host:
            return "https://manhuato.com/"
        
        # Remove common CDN subdomains to construct a clean fallback base domain referer
        clean_host = host
        for prefix in ["cdn.", "img.", "images.", "pic.", "pics.", "static.", "assets.", "media.", "uploads.", "files.", "storage."]:
            if clean_host.startswith(prefix):
                clean_host = clean_host[len(prefix):]
                break
        return f"{parsed.scheme}://{clean_host}/"
    except Exception:
        return "https://www.webtoons.com/"


async def fetch_with_retry(
    url: str,
    headers: dict,
    retries: int = PROXY_MAX_RETRIES,
    base_delay: float = PROXY_RETRY_BASE_SEC
) -> httpx.Response:
    """Fetch with exponential back-off retry on 5xx or network errors."""
    last_err = None
    for attempt in range(retries):
        try:
            async with httpx.AsyncClient(follow_redirects=True, timeout=30.0) as client:
                resp = await client.get(url, headers=headers)
                
                # Retry on 5xx
                if resp.status_code >= 500 and attempt < retries - 1:
                    delay = base_delay * (2 ** attempt)
                    logger.warning(
                        f"[Proxy] Retry {attempt + 1}/{retries} | "
                        f"status {resp.status_code} — waiting {delay:.2f}s"
                    )
                    await asyncio.sleep(delay)
                    continue
                return resp
        except Exception as e:
            last_err = e
            if attempt < retries - 1:
                delay = base_delay * (2 ** attempt)
                logger.warning(
                    f"[Proxy] Network error (retry {attempt + 1}/{retries}): {e} — waiting {delay:.2f}s"
                )
                await asyncio.sleep(delay)
                
    raise last_err or RuntimeError("Max retries reached")


# ─── Routes ───────────────────────────────────────────────────────────────────

@router.get("/proxy-image", summary="Spoofed referrer image bypass proxy")
async def proxy_image(
    request: Request,
    url: str = Query(..., description="Target image URL to fetch")
):
    import asyncio  # Re-import to guarantee availability in route task thread
    
    start_time = time.time()
    
    # Unwrap any double-proxied URLs
    fetch_url = url
    if "/api/proxy-image" in fetch_url:
        parsed = urlparse(fetch_url)
        query = parse_qs(parsed.query)
        if "url" in query:
            fetch_url = query["url"][0]

    # Validate URL format and handle local/internal URLs by redirecting directly
    try:
        parsed_url = urlparse(fetch_url)
        is_local = False
        if not parsed_url.scheme:
            is_local = fetch_url.startswith("/")
        else:
            host = (parsed_url.hostname or "").lower()
            if host in ("localhost", "127.0.0.1") or parsed_url.path.startswith("/api/"):
                is_local = True

        if is_local:
            redirect_url = parsed_url.path
            if parsed_url.query:
                redirect_url += f"?{parsed_url.query}"
            from fastapi.responses import RedirectResponse
            logger.info(f"[Proxy] Redirecting local/internal URL directly to: {redirect_url}")
            return RedirectResponse(url=redirect_url)

        if parsed_url.scheme not in ('http', 'https'):
            raise HTTPException(status_code=400, detail="Only HTTP/HTTPS URLs are supported.")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Invalid URL format: {e}")

    # Cache lookup
    cache_key = hashlib.md5(fetch_url.encode('utf-8')).hexdigest()
    cached = proxy_cache.get(cache_key)

    if cached:
        # Check Client Conditional ETag (304 Not Modified)
        client_etag = request.headers.get("if-none-match")
        if client_etag == cached["etag"]:
            elapsed = int((time.time() - start_time) * 1000)
            logger.info(f"[Proxy] 304 CACHE HIT | {fetch_url[:55]} ({elapsed}ms)")
            return Response(status_code=304)

        elapsed = int((time.time() - start_time) * 1000)
        logger.info(f"[Proxy] 200 CACHE HIT | {fetch_url[:55]} ({cached['size'] / 1024:.1f}KB) ({elapsed}ms)")
        
        return Response(
            content=cached["data"],
            media_type=cached["contentType"],
            headers={
                "ETag": cached["etag"],
                "X-Cache": "HIT",
                "Cache-Control": "public, max-age=31536000, immutable",
                "X-Proxy-Size-KB": f"{cached['size'] / 1024:.1f}"
            }
        )

    # Remote fetch
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
            'Referer':    spoof_referer(fetch_url),
            'Accept':     'image/webp,image/avif,image/jpeg,image/png,image/*,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.9',
            'Accept-Encoding': 'gzip, deflate, br',
        }

        response = await fetch_with_retry(fetch_url, headers)

        if response.status_code != 200:
            logger.warning(f"[Proxy] Upstream error {response.status_code} | {fetch_url[:60]}")
            return Response(
                status_code=response.status_code,
                content=json.dumps({"success": False, "error": f"Upstream returned {response.status_code}", "url": fetch_url}),
                media_type="application/json"
            )

        # Validate content type
        content_type = response.headers.get("Content-Type", "image/jpeg")
        if not content_type.startswith("image/") and "octet-stream" not in content_type:
            logger.warning(f"[Proxy] Blocked non-image response: {content_type} | {fetch_url[:50]}")
            raise HTTPException(
                status_code=415,
                detail=f"Upstream returned non-image content type: {content_type}"
            )

        # Read binary data
        buffer = response.content

        # Size guard
        if len(buffer) > MAX_PROXY_SIZE:
            logger.warning(f"[Proxy] Blocked oversized response: {len(buffer) / 1024 / 1024:.1f}MB > {MAX_PROXY_SIZE_MB}MB limit")
            raise HTTPException(
                status_code=413,
                detail=f"Image exceeds maximum proxy size of {MAX_PROXY_SIZE_MB}MB"
            )

        etag = make_etag(buffer)
        clean_content_type = content_type.split(";")[0].strip()

        # Cache saving
        proxy_cache.set(cache_key, {
            "data": buffer,
            "contentType": clean_content_type,
            "etag": etag,
            "size": len(buffer),
            "fetchedAt": time.time()
        })

        elapsed = int((time.time() - start_time) * 1000)
        logger.info(f"[Proxy] 200 FETCH | {fetch_url[:55]} ({len(buffer) / 1024:.1f}KB) ({elapsed}ms)")

        return Response(
            content=buffer,
            media_type=clean_content_type,
            headers={
                "ETag": etag,
                "X-Cache": "MISS",
                "Cache-Control": "public, max-age=31536000, immutable",
                "X-Proxy-Source": parsed_url.hostname or "",
                "X-Proxy-Size-KB": f"{len(buffer) / 1024:.1f}"
            }
        )

    except HTTPException:
        raise
    except Exception as e:
        elapsed = int((time.time() - start_time) * 1000)
        logger.error(f"[Proxy] ERROR | {fetch_url[:60]} — {e} ({elapsed}ms)")
        raise HTTPException(
            status_code=500,
            detail=f"Proxy fetch failed: {e}"
        )


@router.get("/proxy-cache/stats", summary="Get proxy cache metrics")
async def get_proxy_cache_stats():
    return {"success": True, **proxy_cache.stats()}


@router.delete("/proxy-cache", summary="Clear proxy cache in-memory entries")
async def clear_proxy_cache():
    size = proxy_cache.size
    proxy_cache.clear()
    logger.info(f"[Proxy] Cache cleared — removed {size} entries")
    return {"success": True, "cleared": size}
